In [106]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn import metrics
from sklearn.cluster import KMeans
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [107]:
data = pd.read_csv("CC GENERAL.csv")

data.head()

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0  C10001    40.900749           0.818182      95.40              0.00   
1  C10002  3202.467416           0.909091       0.00              0.00   
2  C10003  2495.148862           1.000000     773.17            773.17   
3  C10004  1666.670542           0.636364    1499.00           1499.00   
4  C10005   817.714335           1.000000      16.00             16.00   

   INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                    95.4      0.000000             0.166667   
1                     0.0   6442.945483             0.000000   
2                     0.0      0.000000             1.000000   
3                     0.0    205.788017             0.083333   
4                     0.0      0.000000             0.083333   

   ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                    0.000000                          0.083333   
1                    0.000000                          0.000000   
2                    1.000000                          0.000000   
3                    0.083333                          0.000000   
4                    0.083333                          0.000000   

   CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                0.000000                 0              2        1000.0   
1                0.250000                 4              0        7000.0   
2                0.000000                 0             12        7500.0   
3                0.083333                 1              1        7500.0   
4                0.000000                 0              1        1200.0   

      PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0   201.802084        139.509787          0.000000      12  
1  4103.032597       1072.340217          0.222222      12  
2   622.066742        627.284787          0.000000      12  
3     0.000000               NaN          0.000000      12  
4   678.334763        244.791237          0.000000      12

In [108]:
data.shape      # fetching attributes and records in given dataset

(8950, 18)

In [109]:
df = data.drop("CUST_ID", axis=1)  #Dropping customer ID

df.isnull().sum()

BALANCE                               0
BALANCE_FREQUENCY                     0
PURCHASES                             0
ONEOFF_PURCHASES                      0
INSTALLMENTS_PURCHASES                0
CASH_ADVANCE                          0
PURCHASES_FREQUENCY                   0
ONEOFF_PURCHASES_FREQUENCY            0
PURCHASES_INSTALLMENTS_FREQUENCY      0
CASH_ADVANCE_FREQUENCY                0
CASH_ADVANCE_TRX                      0
PURCHASES_TRX                         0
CREDIT_LIMIT                          1
PAYMENTS                              0
MINIMUM_PAYMENTS                    313
PRC_FULL_PAYMENT                      0
TENURE                                0
dtype: int64

In [110]:
df = df.fillna(df.mean())   #Filling with mean in place of null

In [111]:
pca = PCA(2)

pca.fit(df)

PCA(n_components=2)

In [112]:
x_pca = pca.fit_transform(df)

df1 = pd.DataFrame(data=x_pca, columns=['principal component 1', 'principal component 2'])

print(df1)         #Printing the 2 components after applying PCA

      principal component 1  principal component 2
0              -4326.383956             921.566884
1               4118.916676           -2432.846347
2               1497.907660           -1997.578692
3               1394.548556           -1488.743450
4              -3743.351874             757.342659
...                     ...                    ...
8945           -4208.357938            1122.443274
8946           -4123.924001             951.683803
8947           -4379.444202             911.504566
8948           -4791.117744            1032.540944
8949           -3623.702749            1555.134769

[8950 rows x 2 columns]


In [113]:
nclusters =2    # Applying K-means algorithm on PCA result

k_mean = KMeans(n_clusters=nclusters)

k_mean.fit(df1)

KMeans(n_clusters=2)

In [114]:
cluster_predict = k_mean.predict(df1)
score = metrics.silhouette_score(df1, cluster_predict)
print("Silhouette score: ",score)         #Calculating the silhouette score PCA + K-means

Silhouette score:  0.598017029765804


In [115]:
scaler = StandardScaler()

x_scaler  = scaler.fit_transform(df)        # applying feature scale to input data

In [116]:
x_pca2 = pca.fit_transform(x_scaler)    # applying PCA on scaled data

k_mean.fit(x_pca2)                      # performing kmeans

KMeans(n_clusters=2)

In [117]:
cluster1_predict = k_mean.predict(x_pca2)
score2 = metrics.silhouette_score(x_pca2, cluster1_predict)
print("Silhouette score: ",score2)         #Getting the silhouette score after applying Scaling + K-means + PCA

Silhouette score:  0.46475556288778747


# As we can see, the silhouette score decreases after scaling the data and then preprocessing it with PCA and applying K-means

Silhouette score (PCA + kmeans) = 0.5980170297658037

Silhouette score (scaling + PCA + kmeans) = 0.46475556288778747